### DISCUSSION TITLE
Cursor input lags

### DISCUSSION THREAD

**Discussion by BenCheung0422** [ppy/osu#27608] [Q&A] [answered]
### Type

Performance

### Bug description

Sometimes when playing with my mouse cursor, the cursor may suddenly stop and continue its relative movements (without teleportation), or with slow movements for a certain span, varies from a few hundreds to a few seconds. From the recordings, comparing the properties between tablet pen (as in #27227) and mouse, it is possible that all of the cases mentioned are related to the drop/loss in input frames. This may be similar to #27488, but I am not sure.

For example, if there is a loss in input frame occurring for tablet pen, the next frame would be the new absolute position corresponding to the position of the pen cursor; if there is a loss in input frame occurring for mouse, the next frame would be the relative movement of the mouse cursor as the previous frame including the data of the cursor movement is lost. Please note that this is just my guess.

Furthermore, from the latest recording without stats overlay, no performance input is logged, which means no slow frames are detected, and no logs can be found around the time, even in the system logs. It is possible that the criteria for incidents might be hard to be spotted out.

The causes of this problem might not be related to GC (as in the two other videos attached with statistics overlay, but it is possible that the overlay causes the lags instead), and may possibly be located by profiler/debugger (possibly internal causes).

### Screenshots or videos

Slow movements (for around a second)

https://github.com/ppy/osu/assets/74168521/d942f850-615e-490b-ae73-94fb234f0cf5

For the video above, I used 2880x1620 DSR for my 1920x1080 monitor recorded with OBS, so that I can put the 2 console windows (smaller) on top on the full-screen game window in output. The font size of the consoles is also changed from 12 (default) to 7 to render more text.

Slow movements (for around 2 seconds; recorded in 2024.302.1)

https://github.com/ppy/osu/assets/74168521/ec2139e2-7618-4fcd-a333-41889e869ea4

Sudden stop (blue circles "1" to "2"; recorded in 2024.302.1)

https://github.com/ppy/osu/assets/74168521/2b300a42-3f08-4ba0-960d-56b6b50129de

### Version

2024.312.1

### Logs

Logs from the latest recording above

[logs-1710333379.zip](https://github.com/ppy/osu/files/14589506/logs-1710333379.zip)

**Comment by EVAST9919** [comment]
https://github.com/ppy/osu-framework/issues/6181

**Comment by frenzibyte** [comment]
> [ppy/osu-framework#6181](https://github.com/ppy/osu-framework/issues/6181)

Not relevant apparently, see https://github.com/ppy/osu/issues/27227#issuecomment-1992515208.

**Comment by peppy** [comment]
Please take the video with the "global statistics" hidden, but ctrl-f11 twice for graphs.

**Comment by BenCheung0422** [comment]
Slow movement for around a second; FPS maxima recorded (in the 60FPS video) within the period: lowest: 934; highest: 1025

https://github.com/ppy/osu/assets/74168521/40662e8a-6eba-42be-8e60-49a2eedcedf6

In addition, I have clipped 9 clips that I have the cursor lag moments rather than with a longer period:

https://github.com/ppy/osu/assets/74168521/66c40952-95e9-413e-898d-508b676f782a

And I tried to take the reference play (with a map of similar difficulty) for comparison and current average situation, but not quite well, I think (made 2 references for average; separated as 2 clips because of file size):

https://github.com/ppy/osu/assets/74168521/20e4c048-a046-4184-b1f5-78e933e1c21b

https://github.com/ppy/osu/assets/74168521/85217c5f-ad03-40a8-a141-2a3c5509dac4

Actually, I can subjectively feel of keyboard not responding well sometimes, but I would assume that to be membrane durability problem. Also, I use a piece of A4 paper for my mouse as a mat, so it is possible that the accuracy might be lower, and some small inaccuracy in position would be expected.

**Reply by peppy** [reply, marked as answer]
Those are showing another app on your system blocking the windows event loop.

You could try to figure which app is doing that, or wait for us to (eventually) fix this via https://github.com/ppy/osu-framework/issues/6181.

**Reply by BenCheung0422** [reply]
After my some small troubleshoots, there might be several apps affecting the input thread. If the gray peaks in the graph means the delays, then according to the amount and lengths of the peaks:
- WhatPulse (obvious)
- Chrome (maybe; not quite obvious; I have dozens of tabs)
- OBS (maybe; not quite obvious)

And it seems like it is not that lag for me today, compared with yesterday (and I felt like it was lagger yesterday than the day before).
Probably when there are some delays in the input thread, some frames are dropped, causing loss of cursor inputs, but I am not sure if it is true. For different circumstances, there might be different causes, but mostly the similar result, loss of frames in cursor inputs. I am guessing that the lags are due to the loss in input frames, and it is currently using the relative movements for mouse cursor, but I am not sure if this assumption is true.
